Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1367,6 +1394,55 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
     return lastInFlowChildBox() == &child;
 }
 
+void RenderBlockFlow::trimFloatBlockEndMargins(LayoutUnit blockFormattingContextInFlowContentHeight)
+{
+    ASSERT(m_floatingObjects && establishesBlockFormattingContext());
+    auto computeFloatMarginAfterStartLocationInBlockFormattingContext = [&](const RenderBox& floatBox) {
+        // Start with the location of the float within its containing block. If this containing
+        // block is the BFC in which we are trimming, then we can use this as the starting location
+        // since this is the float's positin within the BFC. If not, add the containing block's location
+        // to the float's location. This new value should be the location of the float relative to this
+        // new containing block. Keep doing this until we find the containing block that establishes
+        // a BFC
+        auto bfcRoot = floatBox.blockFormattingContextRoot();
+        ASSERT(bfcRoot);
+
+        auto startLocation = floatBox.localToAbsolute();
+        auto containingBlock = floatBox.containingBlock();
+        while (!containingBlock->establishesBlockFormattingContext()) {
+            startLocation += containingBlock->localToAbsolute();
+            containingBlock = containingBlock->containingBlock();
+        }
+        if (bfcRoot->isHorizontalWritingMode())
+            return LayoutUnit { startLocation.y() + bfcRoot->logicalHeightForChild(floatBox) };
+        return LayoutUnit { startLocation.x() + bfcRoot->logicalWidthForChild(floatBox) };
+    };
+    for (auto& floatingObject : m_floatingObjects->set()) {
+        auto& floatBox = floatingObject->renderer();
+        if (!floatBox.containingBlock()->style().marginTrim().contains(MarginTrimType::BlockEnd) || !floatBox.marginAfter())
```

Comment:
what happens to negative margins?

Reply:
FloatingObject has a setMarginOffset() function. Do you know if these need to be in sync?

Reply:
I was having a hard time finding a test case where negative margins for a float increased the height of the BFC. Is that possible? For now I changed it to skip the float if the margin is <= 0